back/llm.py

```diff
@@ -0,0 +1,174 @@
+                messages=[
+                    {   # Keep static content like instructions in beginning to allow prompt caching
+                        "role": "developer", 
+                        "content": """You are an expert in writing headlines in the style of different authors.
+                                        Rewrite the the headline provided to you in the style of the given author, while keeping the same meaning.
+                                        Use the article body for context. Output only the transformed headline, nothing else."""
+                    },
+                    {
+                        "role": "user", 
+                        "content":  f"""Original headline: {headline}
+
+                                        Author style to mimic: {author}
+
+                                        Article body:
+                                        {body}"""
+                    } 
+                ]
+            )
+
+            return response.choices[0].message.content.strip()
+        except Exception as e:
+            print(f"Error in transform_headline: {str(e)}")
+            return headline  # Return original headline if transformation fails
+        
+# Factory for creating LLM provider instances
+class LLMFactory:
+    # Class variable to store provider mapping
+    _providers: Dict[LLMProvider, Type[LLM]] = {
+        LLMProvider.OPENAI: OpenAILLM,
+        LLMProvider.TEST: TestLLM
+    }
+        
+    # Class variable to store instances
+    _instances: Dict[LLMProvider, LLM] = {}
+
+    @classmethod
+    def create(cls, provider: LLMProvider, reuse_instance: bool = True) -> LLM:
+        """
+        Create (or retrieve) an instance of the specified LLM provider.
+        
+        Args:
+            provider: The type of LLM provider to create
+            reuse_instance: If True, reuse existing instance if available
+        
+        Returns:
+            An instance of the requested LLM provider
+        
+        Raises:
+            ValueError: If the requested provider is not registered
+        """
+        if provider not in cls._providers:
+            raise ValueError(f"Unsupported LLM provider: {provider}")
+
+        # Reuse existing instance if requested and available
+        if reuse_instance and provider in cls._instances:
+            return cls._instances[provider]
+
+        # Create new instance
+        instance = cls._providers[provider]()
+        
+        # Store instance if reuse is enabled
+        if reuse_instance:
+            cls._instances[provider] = instance
+
+        return instance
+    
+class HeadlineTransformService:
+    def __init__(self):
+        self.factory = LLMFactory()
+        
+    async def transform_headline(
+        self, 
+        headline: str, 
+        author: str, 
+        body: str, 
+        provider: LLMProvider,
+        fallback_provider: Optional[LLMProvider] = None
+    ) -> tuple[str, LLMProvider]:
```

Comment:
An alternative design for this would be using dependency injection. You could pass in the `provider` and `fallback_provider` to the constructor. Then `transform_headline` would look like:
```
async def transform_headline(
        self, 
        headline: str, 
        author: str, 
        body: str, 
        provider: LLMProvider = self.provider
):
  try:
    ...
  except:
    return self.transform_headline(..., provider=self.transform_headline)
```
This also removes the coupling between `HeadlineTranformService` and `LLMFactory` and just makes it dependent on an interface that, say, a test class can pass in for mocking a call to OpenAI

Reply:
I guess you'd need to check if the `provider` is `self.fallback_provider` in this example and `raise` in that case. Another approach could be to pass in a `FallbackLLM` service that implements the LLM interface that handles fallback logic